File: Platform/Workspace/CSharp/Allors.Workspace/ISessionObject.cs
// <copyright file="SessionObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace
{
    using Allors.Protocol.Remote.Push;
    using Allors.Workspace.Meta;

    public interface ISessionObject
    {
        long Id { get; }

        long? NewId { get; set; }

        long? Version { get; }

        IClass ObjectType { get; }

        ISession Session { get; }

        IWorkspaceObject WorkspaceObject { get; set; }

        bool HasChanges { get; }

        bool CanRead(IRoleType roleType);

        bool CanWrite(IRoleType roleType);

        bool Exist(IRoleType roleType);

        object Get(IRoleType roleType);

        void Set(IRoleType roleType, object value);

        void Add(IRoleType roleType, ISessionObject value);

        void Remove(IRoleType roleType, ISessionObject value);

        PushRequestObject Save();

        PushRequestNewObject SaveNew();

        void Reset();
    }
}

File: Platform/Workspace/Typescript/framework/workspace/ISessionObject.ts
import { ObjectType } from '../meta/ObjectType';
import { ISession } from './Session';
import { IWorkspaceObject } from './WorkspaceObject';
import { RoleType } from '../meta/RoleType';
import { MethodType } from '../meta/MethodType';
import { OperandType } from '../meta/OperandType';
import { Operations } from '../protocol/Operations';
import { AssociationType } from '../meta/AssociationType';
import { PushRequestObject } from '../protocol/push/PushRequestObject';
import { PushRequestNewObject } from '../protocol/push/PushRequestNewObject';

export interface IObject {
  id: string;
  objectType: ObjectType;
}

export interface ISessionObject extends IObject {
  objectType: ObjectType;
  id: string;
  newId?: string;
  version?: string;

  isNew: boolean;

  session: ISession;
  workspaceObject?: IWorkspaceObject;

  hasChanges: boolean;

  canRead(roleType: RoleType): boolean | undefined;
  canWrite(roleTyp: RoleType): boolean | undefined;
  canExecute(methodType: MethodType): boolean | undefined;
  isPermited(operandType: OperandType, operation: Operations): boolean | undefined;

  get(roleType: RoleType): any;
  set(roleType: RoleType, value: any): void;
  add(roleType: RoleType, value: any): void;
  remove(roleType: RoleType, value: any): void;

  getAssociation(associationType: AssociationType): any;

  save(): PushRequestObject | undefined;
  saveNew(): PushRequestNewObject;
  reset(): void;
}
